### PR TITLE
Add retry as failure handling

### PR DESCRIPTION
**Pull request by gluke77**
This change is [Reviewable](https://reviewable.io/reviews/mirantis/k8s-appcontroller/174)

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -311,41 +311,61 @@ func createResources(toCreate chan *ScheduledResource, finished chan string, ccL
...
-							break
-						}
+			attempts := 3
+
+			for attemptNo := 1; attemptNo <= attempts; attemptNo++ {
+				var err error
+
+				if attemptNo > 1 {
+					log.Printf("Trying to delete resource %s after previous unsuccessful attempt", r.Key())
+					err = r.Delete()
```
Comment (dshulyak) [inline pkg/scheduler/scheduler.go]: what error we will prevent by recreating exactly same resource?

Comment (gluke77) [inline pkg/scheduler/scheduler.go]: It's still WIP )

Comment (gluke77) [inline pkg/scheduler/scheduler.go]: Actually, whatever error end user would like, providing retry parameter. E.g. temporarily network issues)

**Comment by nebril** [comment]
Reviewed 20 of 20 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[examples/test/create.sh, line 1 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174#-Kaq2hSPU1QmwtEkOEXz:-Kaq2hSQHhBfnn5rJ_xB:b-rknko4) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/47f10ef6ada5aaa2b40722ab4349ce76909d031e/examples/test/create.sh#L1)):*
> ```Shell
> #!/bin/bash
> ```

What is this example testing exactly?

---

*[pkg/resources/common.go, line 167 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174#-Kaq36kTD-cfKfXA8erF:-Kaq36kTD-cfKfXA8erG:b-k08ge1) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/47f10ef6ada5aaa2b40722ab4349ce76909d031e/pkg/resources/common.go#L167)):*
> ```Go
> 	}
> 
> 	intVal, ok := value.(float64)
> ```

Do we really want to implicitly round floats to ints?

---

*[pkg/scheduler/scheduler.go, line 328 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174#-Kaq3qWSZE-0N8je40Pi:-Kaq3qWSZE-0N8je40Pj:b-9oqi9q) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/47f10ef6ada5aaa2b40722ab4349ce76909d031e/pkg/scheduler/scheduler.go#L328)):*
> ```Go
> 				}
> 
> 				log.Printf("Creating resource %s, attempt %d of %d", r.Key(), attemptNo, attempts)
> ```

It seems like sth funky happened with indentation here? Shouldn't this code be on the same level the `if err != nil` is?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174)*

**Comment by gluke77** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[examples/test/create.sh, line 1 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174#-Kaq2hSPU1QmwtEkOEXz:-Kaq8MrF_H2ABHMygwSG:b-h6x8lg) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/47f10ef6ada5aaa2b40722ab4349ce76909d031e/examples/test/create.sh#L1)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

What is this example testing exactly?
</blockquote></details>

oh, tnx, it's just a leftover from manual testing. will remove it

---

*[pkg/resources/common.go, line 167 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174#-Kaq36kTD-cfKfXA8erF:-Kaq8ZCEAGX5xBbSkpEI:b7uiylv) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/47f10ef6ada5aaa2b40722ab4349ce76909d031e/pkg/resources/common.go#L167)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

Do we really want to implicitly round floats to ints?
</blockquote></details>

Yeah. Json seems to hold floats always. And the func returns ints. I expect more helpers for different types here, but in the case I do want int.

---

*[pkg/scheduler/scheduler.go, line 328 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174#-Kaq3qWSZE-0N8je40Pi:-Kaq8vxRoEzLr6futV4_:bx1fujh) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/47f10ef6ada5aaa2b40722ab4349ce76909d031e/pkg/scheduler/scheduler.go#L328)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

It seems like sth funky happened with indentation here? Shouldn't this code be on the same level the `if err != nil` is?
</blockquote></details>

nope, it logs even the 1st attempt as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174)*

**Comment by nebril** [comment]
Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174)*

**Comment by nebril** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174#-:-KaqzohDg1Z6t2f2ChRn:bnfp4nl)*

**Comment by pigmej** [comment]
:lgtm:

---

Reviewed 17 of 20 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/174#-:-Kaw1VxScMaPStvTHnw1:bnfp4nl)*

**Pull request merged**